flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeSchedulerITCase.java

```diff
@@ -0,0 +1,245 @@
+        configuration.set(
+                JobManagerOptions.SCHEDULER, JobManagerOptions.SchedulerType.AdaptiveBatch);
+        configuration.set(JobManagerOptions.SPECULATIVE_ENABLED, true);
+        // for testing, does not block node by default
+        if (!configuration.contains(JobManagerOptions.BLOCK_SLOW_NODE_DURATION)) {
+            configuration.set(JobManagerOptions.BLOCK_SLOW_NODE_DURATION, Duration.ZERO);
+        }
+        configuration.set(SlowTaskDetectorOptions.EXECUTION_TIME_BASELINE_MULTIPLIER, 1.0);
+        configuration.set(SlowTaskDetectorOptions.EXECUTION_TIME_BASELINE_RATIO, 0.2);
+        configuration.set(
+                SlowTaskDetectorOptions.EXECUTION_TIME_BASELINE_LOWER_BOUND, Duration.ofMillis(0));
+
+        // for adaptive parallelism
+        configuration.set(
+                JobManagerOptions.ADAPTIVE_BATCH_SCHEDULER_DEFAULT_SOURCE_PARALLELISM,
+                MAX_PARALLELISM);
+        configuration.set(JobManagerOptions.ADAPTIVE_BATCH_SCHEDULER_MIN_PARALLELISM, 1);
+        configuration.set(
+                JobManagerOptions.ADAPTIVE_BATCH_SCHEDULER_MAX_PARALLELISM, MAX_PARALLELISM);
+        configuration.set(
+                JobManagerOptions.ADAPTIVE_BATCH_SCHEDULER_AVG_DATA_VOLUME_PER_TASK,
+                MemorySize.parse("150kb"));
+
+        final StreamExecutionEnvironment env =
+                StreamExecutionEnvironment.createLocalEnvironment(configuration);
+        env.setRuntimeMode(RuntimeExecutionMode.BATCH);
+        env.setParallelism(-1);
+
+        final DataStream<Long> source =
+                env.fromSequence(0, NUMBERS_TO_PRODUCE - 1)
+                        .setParallelism(parallelism)
+                        .name("source")
+                        .slotSharingGroup("group1");
+
+        final DataStream<Long> map =
+                source.rebalance()
+                        .map(new NumberCounterMap())
+                        .setParallelism(parallelism)
+                        .name("map")
+                        .slotSharingGroup("group2");
+
+        map.rebalance()
+                .addSink(new NumberCounterSink())
+                .setParallelism(parallelism)
+                .name("sink")
+                .slotSharingGroup("group3");
+
+        return env.executeAsync();
+    }
+
+    private void waitUntilJobArchived() throws InterruptedException {
+        while (temporaryFolder.getRoot().toFile().listFiles().length < 1) {
+            Thread.sleep(1000);
+        }
+    }
+
+    private static class NumberCounterMap extends RichMapFunction<Long, Long> {
+        private static final AtomicInteger toFailCounter = new AtomicInteger(0);
+
+        private boolean hasWaited = false;
+
+        @Override
+        public Long map(Long value) throws Exception {
+            if (toFailCounter.decrementAndGet() >= 0) {
+                throw new Exception("Forced failure for testing");
+            }
+
+            if (!hasWaited) {
```

Comment:
We can simplify it as below, the `hasWaited` can be removed then.
```
           if (isSlowTask()) {
                Thread.sleep(1_000_000);
            }
```

Reply:
Fixed